urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/interp/ParserThread.java

```diff
@@ -0,0 +1,110 @@
+	private Scanner scanner;      public Scanner scanner() { return(scanner); }
+	private Reader in;            public Reader in() { return(in); }
+	private Symbol pendingToken;
+	private PTfactory pf;         public PTfactory pf() { return(pf); } 
+	private SymbolName sname;     public SymbolName sname() { return(sname); }
+	//this is not a atom opeartion, so it's not a very good solution
+	//public static boolean flag = true;
+	private BlockingQueue<PTeval> queue; public BlockingQueue<PTeval> queue(){return queue;}
+	
+	public ParserThread(Reader _in, PTfactory _pf, BlockingQueue<PTeval> q, SymbolName _sname) {
+		super();
+		if(_in == null) _in = new InputStreamReader(System.in);
+		if(_pf == null) _pf = new PTnodeFactory();
+		if(_sname == null) _sname = new SymbolName();
+		in = _in;
+		pendingToken = null;
+		pf = _pf;
+		queue = q;
+		sname = _sname;
+		scanner = new Scanner(in, pf, sname);
+	}
+	
+	public Parse create(Reader in) {
+		return(new ParserThread(in, pf,queue, sname));
+	}
+	
+	
+	// override parser.scan to use scanner
+	public Symbol scan() throws java.io.IOException {
+		Symbol s;
+		if(pendingToken != null) {
+			s = pendingToken;
+			pendingToken = null;
+		}  else {
+			PTnode p;
+			p = (PTnode)(scanner.next_token().value);//change after %cup
+			s = new Symbol(p.key(), p);
+		}
+		if(in instanceof PromptingReader)
+			((PromptingReader)(in)).setPrompt(null);
+		return(s);
+	}
+	
+	// override calc.syntax_error(Symbol) to do nothing.
+	//   We'll print the error message ourself when we catch the
+	//   Exception thrown by our version of unrecovered_syntax_error().
+	public void syntax_error(Symbol cur_token) {
+		report_error("Syntax error: "+cur_token.toString(),null);
+	}
+	
+	// override calc.unrecovered_syntax_error(Symbol) to silently throw
+	//   an EvalException
+	public void unrecovered_syntax_error(Symbol cur_token)throws EvalException {
+		PTnode tkn = (PTnode)(cur_token.value);
+		if(tkn.key() == Sym.LEX_ERROR)
+			throw new EvalException("lexical error: " + tkn.text());    	
+		else 
+			throw new EvalException("syntax error"+"\n Debug info:"+tkn.toString());
+	}
+	
+	// override error_recovery(boolean) to do nothing (the method in
+	//   lr_parser reads three tokens of look-ahead, which messes up
+	//   what we want to do in unrecovered_syntax_error().
+	protected boolean error_recovery(boolean debug) {
+		return(false);  // just fail
+	}
+	public void report_error(String message, Object info){
+		System.out.println("report_error: "+message.toString());
+	}
+	public void report_fatal_error(String message, Object info){
+		report_error(message,info);
+		done_parsing();
+	}
+	
+	/* override init_actions() call our version */
+	protected void init_actions() { init_actions_2(pf,queue); }
+	
+	
+	public void run(){
+		try{
+			parse();
+			//flag =false;
+			//we add a special PTeval instead.
+			try{
+				queue.put(BasicPTeval.voidPTeval);
+			}catch(InterruptedException ee){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)